discovery/acl.go

```diff
@@ -0,0 +1,69 @@
+package discovery
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul-server-connection-manager/internal/consul-proto/pbacl"
+	"github.com/hashicorp/go-hclog"
+)
+
+type ACLs interface {
+	Login(context.Context) (string, error)
+	Logout(context.Context) error
+}
+
+type defaultACLs struct {
+	watcher *Watcher
+	log     hclog.Logger
+}
+
+var _ ACLs = (*defaultACLs)(nil)
+
+func newDefaultACLs(watcher *Watcher, log hclog.Logger) *defaultACLs {
+	return &defaultACLs{watcher: watcher, log: log}
+}
+
+func (a *defaultACLs) Login(ctx context.Context) (string, error) {
+	if a.watcher.config.Credentials.Type != CredentialsTypeLogin {
+		return "", nil
+	}
+
+	cfg := a.watcher.config.Credentials.Login
+
+	client := pbacl.NewACLServiceClient(a.watcher.conn)
+	resp, err := client.Login(ctx, &pbacl.LoginRequest{
+		AuthMethod:  cfg.AuthMethod,
+		BearerToken: cfg.BearerToken,
+		Meta:        cfg.Meta,
+		Namespace:   cfg.Namespace,
+		Partition:   cfg.Partition,
+		Datacenter:  cfg.Datacenter,
+	})
+	if err != nil {
+		a.log.Error("auth method login failed", "error", err)
+		return "", err
+	}
+
+	tok := resp.GetToken()
+	a.log.Info("auth method login successful", "accessor-id", tok.AccessorId)
+	// TODO: do the acl token replication workaround here?
+	//       https://github.com/hashicorp/consul-k8s/pull/887
```

Comment:
Great call out. Instead of here, maybe calling it from the watcher.connect after login and before fetching dp features? 
Also, what were you thinking about which call to make to validate the token has replicated over? Is it something we can retry on while fetching dp features?

Reply:
I haven't really thought it through, but I don't think it's something we have to do immediately.

So, this is a problem only for requests made in stale consistency mode. And I think only when a subsequent request reaches a server follower that is different than the server that responded to the ACL login request. And only for a short period after the ACL token login (basically however long Raft replication takes, which is something like 100ms typically).

Our case is a little different, too. When servers are not behind a load balancer, then it shouldn't be an issue because we connect to specific server and stay connected to that server after a successful ACL login. And we know when we switch servers.

It may crop up if the servers are behind a load balancer, like if the load balancer round robins gRPC messages on a single connection (is this super common?).

Reply:
Yea leaving a TODO for now should be fine and we can revisit based on testing more. 

> And I think only when a subsequent request reaches a server follower that is different than the server that responded to the ACL login request.

^ this is applicable to consul-dataplane too, yea?
In case we connect to a specific server that is a follower, the ACL login gets forwarded to the leader server. And then we immediately make a get supported features call on the follower - I was thinking this is where it would impact the request in case it was made before raft replication finishes.

Reply:
> the [ACL login gets forwarded to the leader server.](https://github.com/hashicorp/consul/blob/main/agent/grpc-external/services/acl/login.go#L34)

Right. I thought a follower would receive the successful ACL login response from the leader, and so it would know that particular token is good to use? And if that's the case, then as long as we hit the same follower, then everything is OK.

I'm not sure about that though.

Reply:
Hmm, but I thought it would not be persisted anywhere for the follower to know. 
@boxofrad can you confirm?

Reply:
Right, I meant the follower would have to look at the ACL login response it receives from a leader. If the ACL login response is successful then the follower could cache the token from in the response as successful for a short time (but not persist the token), rather than waiting for Raft replication from the leader to occur.

Reply:
That'd be a really useful optimization but isn't something we do at the moment, unfortunately.

Currently, we forward the request to the leader or federated DC at the beginning of the RPC handler, and just pass the response back to the client without inspecting it or anything ([source](https://github.com/hashicorp/consul/blob/0d9ae52643a1f75556576a8ca7f674d6efd78588/agent/grpc-external/services/acl/login.go#L34-L43)).

So yeah, you'll need to wait for Raft replication; which means clients might get an `Unauthenticated` response if making the features call straight away.

As an aside: it'd be cool to offer a "read your writes" consistency mode for this kind of thing where the RPC handler blocks until replication is complete.

Reply:
For future reference, per Slack, we'll address this more properly server-side so we don't need the client side waiting / workarounds for now.

In the meantime, I hardcoded a 100ms sleep after the login which seems harmless, and might help avoid some cases of this problem.